Auth0 User Management API upgrade to v2

landonreed opened the pull request:
### Checklist

- [x] Appropriate branch selected _(all PRs must first be merged to `dev` before they can be merged to `master`)_
- [x] Any modified or new methods or classes have helpful JavaDoc and code is thoroughly commented
- [x] The description lists all applicable issues this PR seeks to resolve
- [x] The description lists any configuration setting(s) that differ from the default settings
- [x] All tests and CI builds passing

### Description

This PR migrates the Auth0 Management API from v1 to v2. Instead of using a static token defined in the config, this instead uses a machine-to-machine client/secret pair to request a new token (or use a cached token) every time it needs to make a request to the Management API. Fixes #183

evansiroky commented on src/main/java/com/conveyal/datatools/manager/auth/Auth0Users.java:
```diff
@@ -119,6 +133,54 @@ private static String doRequest(URI uri) {
+            final String audienceUrl = builder.setPath(API_PATH + "/").build().toString();
+            URI uri = builder.setPath("/oauth/token").build();
+            LOG.info(uri.toString());
+            LOG.info(audienceUrl);
```
These logging statements can probably be removed.

evansiroky commented on src/main/java/com/conveyal/datatools/manager/auth/Auth0Users.java:
```diff
@@ -119,6 +133,54 @@ private static String doRequest(URI uri) {
+            post.setEntity(new UrlEncodedFormEntity(urlParameters));
+            HttpResponse response = client.execute(post);
+            int code = response.getStatusLine().getStatusCode();
+            if (code >= 300) LOG.error("Could not get Auth0 API token");
```
I feel like there are a lot of different things that could go wrong in this method that don't have helpful error messages. Like if Auth0 somehow changes their access token response body, this method may begin returning null and there'd be some cryptic stack trace. It would be nice to see each of these exceptions encapsulated within a try/catch block that has a helpful error message and then either immediately returns null or re-throws the exception to a higher level.

evansiroky commented on src/main/java/com/conveyal/datatools/manager/auth/Auth0Users.java:
```diff
@@ -78,7 +92,7 @@ private static String doRequest(URI uri) {
         HttpGet request = new HttpGet(uri);
 
-        request.addHeader("Authorization", "Bearer " + AUTH0_API_TOKEN);
+        request.addHeader("Authorization", "Bearer " + getApiToken());
```
when the return of `getApiToken()` is null this whole request will still continue when it actually should probably immediately stopped since there isn't a valid API token.

evansiroky commented on src/main/java/com/conveyal/datatools/manager/auth/Auth0Users.java:
```diff
@@ -165,20 +169,28 @@ private static String getApiToken() {
+                LOG.error("Could not get Auth0 API token {}", responseString);
+                throw new IllegalStateException("Bad response for Auth0 token");
+            }
         } catch (IllegalStateException | URISyntaxException | IOException e) {
```
Is IOException still needed here?

landonreed replied:
Yea, it's thrown by the few lines pertaining to the request/response entity in this block.

landonreed commented:
@evansiroky, I have added a test case for expected failure of update user and fixed a number of cases in `UserController` where the old API token was still being used.

evansiroky commented on src/main/java/com/conveyal/datatools/manager/controllers/api/UserController.java:
```diff
@@ -73,7 +72,7 @@
     private static String getUser(Request req, Response res) {
         HttpGet request = new HttpGet(getUserIdUrl(req));
-        setHeaders(request);
+        setHeaders(req, request);
```
Can we rename `request` to `getUserRequest` to avoid having two very similarly named variables?

evansiroky commented on src/main/java/com/conveyal/datatools/manager/controllers/api/UserController.java:
```diff
@@ -147,7 +146,7 @@ private static int getUserCount(Request req, Response res) {
     private static String createPublicUser(Request req, Response res) {
         HttpPost request = new HttpPost(baseUsersUrl);
-        setHeaders(request);
+        setHeaders(req, request);
```
Variable names too similar.

evansiroky commented on src/main/java/com/conveyal/datatools/manager/controllers/api/UserController.java:
```diff
@@ -166,7 +165,7 @@ private static String createPublicUser(Request req, Response res) {
     private static String createUser(Request req, Response res) {
         HttpPost request = new HttpPost(baseUsersUrl);
-        setHeaders(request);
+        setHeaders(req, request);
```
Variable names too similar.

evansiroky commented on src/main/java/com/conveyal/datatools/manager/controllers/api/UserController.java:
```diff
@@ -188,14 +187,14 @@ private static String updateUser(Request req, Response res) {
 
         HttpPatch request = new HttpPatch(getUserIdUrl(req));
-        setHeaders(request);
+        setHeaders(req, request);
```
Variable names too similar.

evansiroky commented on src/main/java/com/conveyal/datatools/manager/controllers/api/UserController.java:
```diff
@@ -218,7 +217,7 @@ private static String updateUser(Request req, Response res) {
     private static Object deleteUser(Request req, Response res) {
         HttpDelete request = new HttpDelete(getUserIdUrl(req));
-        setHeaders(request);
+        setHeaders(req, request);
```
Variable names too similar.

evansiroky commented on src/test/java/com/conveyal/datatools/manager/controllers/api/UserControllerTest.java:
```diff
@@ -254,6 +268,81 @@ public void canDeleteUser() throws IOException {
+                .asString()
+        );
+        // Assert that update response matches snapshot (failure message).
+        assertThat(updateUserResponse, matchesSnapshot());
```
I don't see the relevant snapshot checked in with this PR.

landonreed commented:
Thanks for the comments, @evansiroky! Just addressed and will merge once the CI passes.

landonreed commented:
:tada: This PR is included in version 3.5.0 :tada:

The release is available on [GitHub release](https://github.com/ibi-group/datatools-server/releases/tag/v3.5.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
